### PR TITLE
Reduce the amount scaled when using mouse-wheel zooming in Chromium-browsers (issue 16325)

### PR DESCRIPTION
**Pull request by Abishek-Jayan**
The deltaMode value apparently defaults to a different value depending on the browser OS combo (as per https://github.com/w3c/pointerevents/issues/592#issuecomment-392648065 ) . In our case I've noticed that for Chromium browsers the delta value always comes out to a value greater than the PIXELS_PER_LINE_SCALE variable. This is what causes the error. So to fix, I just put a check there and set the correct tick value accordingly. Ive tested in both Edge and Brave browsers and it seems to be working fine both in desktop view and mobile view. 

This is a fix for #16325

### CODE REVIEW

File: web/app.js [hunk 1]
```diff
@@ -2780,12 +2780,14 @@ function webViewerWheel(evt) {
...
-        );
+        ticks =
+          Math.abs(delta) > PIXELS_PER_LINE_SCALE
+            ? Math.sign(delta)
+            : PDFViewerApplication._accumulateTicks(
+                delta / PIXELS_PER_LINE_SCALE,
+                "_wheelUnusedTicks"
+              );
       }
-
```
Comment (Snuffleupagus) [inline web/app.js:2788]: This is an unrelated change, which should be reverted.

**Review by Snuffleupagus — CHANGES_REQUESTED** [review]
Sorry, but I still think this feels risky and it could easily break some other configuration (and thus require follow-up patches); hence I'm unfortunately not comfortable approving this.

---

The PR title mentions scrolling, which seems quite confusing since the issue is regarding mouse-wheel zooming. (Also, the title should use the words `(issue 16325)` and *not* `( bug #16325 )` since the original issue was filed in GitHub and not Bugzilla.)

The *commit message* should have the same title as the PR, and then also contain all of the contents from https://github.com/mozilla/pdf.js/pull/16596#issue-1772688295.

Finally, please note https://github.com/mozilla/pdf.js/wiki/Squashing-Commits since you've got an unrelated commit included.

**Comment by Abishek-Jayan** [comment]
@Snuffleupagus hmm there is one thing i dont understand... in 2782 the variable PIXELS_PER_LINE_SCALE is set as 30? why? If im correct in assuming that that variable represents the number of pixels in one line, couldn't it change based on the size of the device or something?

**Comment by Abishek-Jayan** [comment]
@Snuffleupagus 
So I have found that rather than messing with the delta values, the simplest way to solve the problem would be to find a clean way to 
detect if the browser was Chromium or not. I found that chromium browsers have a property called windows.chrome which on console logging, 
looks something like this:

{
        app: {
            isInstalled: false,
            InstallState: {
                DISABLED: disabled,
                INSTALLED: installed,
                NOT_INSTALLED: not_installed
            },
            RunningState: {
                CANNOT_RUN: cannot_run,
                READY_TO_RUN: ready_to_run,
                RUNNING: running
            }
        },
        appPinningPrivate: {}
}

Whereas in firefox, logging that property returns undefined. I had searched to see if there were ways that the window.chrome property could be faked (since the navigator.userAgent could be faked) but I couldnt find anything regarding that. Whats more, Edge, Opera, Brave and Chrome all support this property as per https://stackoverflow.com/a/13348618

So the current fix just adds the windows.chrome check to the if condition at line 2762.

I had tested this functionality myself on all these browsers and they work as intended in both desktop and mobile versions.

File: web/app.js [hunk 2]
```diff
@@ -2761,7 +2761,8 @@ function webViewerWheel(evt) {
       let ticks = 0;
       if (
         deltaMode === WheelEvent.DOM_DELTA_LINE ||
-        deltaMode === WheelEvent.DOM_DELTA_PAGE
+        deltaMode === WheelEvent.DOM_DELTA_PAGE ||
+        window.chrome
       ) {
```
Comment (Snuffleupagus) [inline web/app.js:2766]: This will bundle useless code in the Firefox PDF Viewer, which we always want to avoid.
```javascript
      if (
        (typeof PDFJSDev !== "undefined" &&
          PDFJSDev.test("GENERIC || CHROME") &&
          window.chrome) ||
        deltaMode === WheelEvent.DOM_DELTA_LINE ||
        deltaMode === WheelEvent.DOM_DELTA_PAGE
      ) {
```

Comment (Abishek-Jayan) [reply web/app.js:2766]: I'm confused, are you saying the window.chrome is useless due to the existence of the PDFJSdev.test command? But clearly the pdfjsdev command isn't doing its job properly if its not detecting the chromium browser

Comment (Snuffleupagus) [reply web/app.js:2766]: I'm simply saying that you'll need to make exactly the changes mentioned above to avoid an unnecessary piece of code ending up in the *built-in* Firefox PDF Viewer.

Comment (Abishek-Jayan) [reply web/app.js:2766]: So I tried out the above change...but then the fix doesn't work anymore. I checked the debugger and that PDFJSDev variable doesnt even exist so maybe thats why the condition fails there itself. In all Chromium browsers it skips past that if statement and goes directly into the corresponding else statement (aka the pixel based calculation), thereby persisting the issue. So that above change doesnt help I think.

Comment (Snuffleupagus) [reply web/app.js:2766]: > So I tried out the above change...but then the fix doesn't work anymore.

Yes it will, once you've built the code.

Comment (Abishek-Jayan) [reply web/app.js]: You were right... I built and tried it out on chrome and it works like a charm. Updated my commit message accordingly.

**Comment by Abishek-Jayan** [comment]
@Snuffleupagus now the current commit should have everything in order

**Comment by Abishek-Jayan** [comment]
@Snuffleupagus  hope everything is correct

**Comment by Snuffleupagus** [comment]
I'm still a little bit worried about this potentially causing issues in some edge-case, or for some users, given that this Chrome-only issue has (as far as I can tell) only been reported once.
@calixteman, @timvandermeij How do you feel about this patch?

**Comment by calixteman** [comment]
> I'm still a little bit worried about this potentially causing issues in some edge-case, or for some users, given that this Chrome-only issue has (as far as I can tell) only been reported once. @calixteman, @timvandermeij How do you feel about this patch?

I don't have windows machine for the moment, but iirc, I already met this issue with chrome on Windows.
I just tested with Chrome on my mac and it behaves correctly.
As far as I can tell this patch will only affect chrome users so I'm fine with it even if I'm pretty sure that it isn't the correct solution.
And as it has been noticed, the value `const PIXELS_PER_LINE_SCALE = 30;` probably deserves to be understood maybe the screen resolution matters here... I don't really know. I'll try to find some time to investigate more deeply.

**Comment by Abishek-Jayan** [comment]
@Snuffleupagus @timvandermeij @calixteman hi, so...is this fix ok?

**Comment by Abishek-Jayan** [comment]
@Snuffleupagus ??

**Comment by Snuffleupagus** [comment]
Closing this for now, since as mentioned in https://github.com/mozilla/pdf.js/pull/16596#issuecomment-1611981899 we want to understand how/why Chrome behaves differently *before* making changes here.
